internal/agent/config.go

```diff
@@ -0,0 +1,129 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package agent
+
+import (
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"github.com/fsnotify/fsnotify"
+	apiv1 "github.com/numaproj-labs/numaplane/api/v1alpha1"
+	"github.com/spf13/viper"
+)
+
+type AgentConfig struct {
+	ClusterName     string `mapstructure:"clusterName"`
+	TimeIntervalSec uint   `mapstructure:"timeIntervalSec"`
+
+	Source Source `mapstructure:"source"`
+}
+
+type Source struct {
+	// optional, apply to GitDefinition
+	KeyValueGenerator *KVGenerator `mapstructure:"keyValueGenerator,omitempty"`
+
+	GitDefinition apiv1.CredentialedGitSource `mapstructure:"gitDefinition"`
+}
+
+type KVGenerator struct {
+	Embedded *apiv1.SingleClusterGenerator `mapstructure:"embedded,omitempty"`
+
+	Reference *apiv1.MultiClusterFileGenerator `mapstructure:"reference,omitempty"`
+}
+
+type ConfigManager struct {
+	config *AgentConfig
+	lock   *sync.RWMutex
+	// revisionIndex increments each time the Config changes so the caller can keep track of when
```

Comment:
Why do we need to keep track of this ourselves?

Reply:
we can avoid reprocessing the file each time by only processing it when it changed